### PR TITLE
1201-add-trt-benchmark-tutorial

### LINKED ISSUE #1201: Add a tutorial of benchmarking trt models
**Is your feature request related to a problem? Please describe.**
Models in MONAI model zoo are going to be optimized by TensorRT and converted to a TensorRT engine.

We'd better to add  a tutorial of how to benchmark this optimized result.

**Describe the solution you'd like**
Add a tutorial to compare the inference time difference between the original module and TensorRT optimized module.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by binliunls**
Fixes #1201 .

### Description
Add a tutorial to compare the inference time difference between the original module and TensorRT optimized module.
In this tutorial, based on the spleen ct segmentation bundle, we will show how to do the things below:
1. Convert a model to a TensorRT engine
2. Benchmark the latency difference between the original model and the trt model
3. How to analyse the latency with Nsight profiler

### Checks

- [ ] Avoid including large-size files in the PR.
- [ ] Clean up long text outputs from code cells in the notebook.
- [ ] For security purposes, please check the contents and remove any sensitive info such as user names and private key.
- [ ] Ensure (1) hyperlinks and markdown anchors are working (2) use relative paths for tutorial repo files (3) put figure and graphs in the `./figure` folder
- [ ] Notebook runs automatically `./runner.sh -t <path to .ipynb file>`

### CODE REVIEW

**Comment by mingxin-zheng** [comment]
Thank you @binliunls for the PR.

The figures lack x/y labels. Please add them in the function (and with the units).

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+    "        timer.start()\n",
+    "\n",
+    "    def completed(self, engine: Engine, timer: CUDATimer) -> None:\n",
+    "        timer.end()\n",
+    "\n",
+    "    def print_results(self) -> None:\n",
+    "        def dict_to_str(results_dict) -> str:\n",
+    "            return \"|\".join(f\"{v:.5f}\" for v in results_dict.values())\n",
+    "\n",
+    "        out_title = \"\"\"\n",
```
Comment (mingxin-zheng) [inline acceleration/TensorRT_inference_acceleration.ipynb:480]: Maybe we can use `pandas` to show the table?

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+    "torch_timer.print_results()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Benchmark the TensorRT fp32 and fp16 models\n",
+    "\n",
+    "Update the config parser with Torch-TensorRT models and benchmark the end to end latency. Since the `POST_PROCESS` and `PREPARE_BATCH` part take a lot of time, the acceleration of model computation doesn't have much effect on the end2end latency."
```
Comment (mingxin-zheng) [inline acceleration/TensorRT_inference_acceleration.ipynb:556]: I would rephrase this sentence:
The `POST_PROCESS` and `PREPARE_BATCH` stages require a considerable amount of time. Although the model forward time is much improved, there is still room for acceleration in reducing the end-to-end latency on this particular MONAI bundle.

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 3]
```diff
@@ -0,0 +1,930 @@
...
+    "\n",
+    "# Benchmark TensorRT models inference"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This tutorial shows how to convert a bundle in the MONAI model zoo to a TensorRT engine-based TorchScript for better performance and benchmark the difference between the PyTorch module and the TensorRT engine-based TorchScript.  \n",
+    "The `spleen_ct_segmentation` bundle is used as the example in this tutorial. The inference latency before and after the covert process will be compared. Feel free to modify the `spleen_ct_segmentation` bundle to other bundles in the MONAI model zoo which support TensorRT conversion. The TensorRT conversion is supported since **MONAI 1.2** with **Torch-TensorRT 1.4.0** and **TensorRT 8.5.3**. Please use the MONAI Docker whose version is greater than 1.2 or ensure the environment meets the min version requirement. The supported bundle names are listed below. Please modify the net_id and input_shape in the tutorial to fit different bundles.\n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:26]: "before and after the covert process" -> "before and after the conversion process"

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 4]
```diff
@@ -0,0 +1,930 @@
...
+    "\n",
+    "import matplotlib.pyplot as plt\n",
+    "import monai\n",
+    "from monai.apps import download_and_extract\n",
+    "from monai.config import print_config\n",
+    "from ignite.engine import Events\n",
+    "from collections import OrderedDict\n",
+    "from monai.engines import IterationEvents\n",
+    "\n",
+    "print(f\"Torch-TensorRT version: {torch_tensorrt.__version__}.\")\n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:89]: Hi @wyli ,

Do you think it's better to include `torch-trt` as part of MONAI optional dependency?
Currently, the main concern is that TensorRT environment is complicated and related to CUDA, cuDNN, TensorRT lib, etc. `pip install torch-trt` may not work well.

Thanks.

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 5]
```diff
@@ -0,0 +1,930 @@
...
+    "train_labels = sorted(glob.glob(os.path.join(data_root, \"labelsTr\", \"*.nii.gz\")))\n",
+    "data_dicts = [{\"image\": image_name, \"label\": label_name} for image_name, label_name in zip(train_images, train_labels)]\n",
+    "train_files, val_files = data_dicts[:-9], data_dicts[-9:]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Export the trt model with API\n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:192]: Export the TensorRT model with API

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 6]
```diff
@@ -0,0 +1,930 @@
...
+   "outputs": [],
+   "source": [
+    "def get_parser(meta_config, inference_config):\n",
+    "    parser = monai.bundle.ConfigParser()\n",
+    "    parser.read_meta(meta_config)\n",
+    "    parser.read_config(inference_config)\n",
+    "    return parser\n",
+    "\n",
+    "\n",
+    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:235]: bundle already defines the `device` in it.
I would suggest to use `device = parser.get_parsed_content("device")` instead.

Thanks.

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 7]
```diff
@@ -0,0 +1,930 @@
...
+    "\n",
+    "parser = get_parser(meta_config, inference_config)\n",
+    "\n",
+    "# If using other bundles, the net_id and spatial_shape may change. Please modify these variables accordingly.\n",
+    "net_id = \"network\"\n",
+    "spatial_shape = (1, 1, 96, 96, 96)\n",
+    "model = parser.get_parsed_content(net_id)\n",
+    "\n",
+    "model.eval()\n",
+    "model.to(device)\n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:250]: No need to move, the parsed `network` object of bundle is already on the target device.

Thanks.

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 8]
```diff
@@ -0,0 +1,930 @@
...
+    "        f\"Average time for the {model_type} model: {average_time}ms.\",\n",
+    "    )\n",
+    "    plt.plot(timeaccumulate)\n",
+    "    plt.xlabel(\"Iteration\")\n",
+    "    plt.ylabel(\"Latency (ms)\")\n",
+    "    plt.title(model_type)\n",
+    "    plt.show(model_type)\n",
+    "\n",
+    "\n",
+    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:355]: No need to define `device` again.

Thanks.

**Review by Nic-Ma — COMMENTED** [review]
Hi @binliunls ,

I see you use `ConfigParser` to handle a bundle config in multiple places of this PR, please update all of them to use `monai.bundle.ConfigWorkflow` instead.
You can refer to: https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_bundle_workflow.py#L204.
And use `ConfigWorkflow initialize/run/finalize` to execute the inference. 
CC @yiheng-wang-nv for any other `BundleWorkflow` related questions.

Thanks.

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 9]
```diff
@@ -0,0 +1,930 @@
...
+    "limitations under the License.\n",
+    "\n",
+    "# Benchmark TensorRT models inference"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This tutorial shows how to convert a bundle in the MONAI model zoo to a TensorRT engine-based TorchScript for better performance and benchmark the difference between the PyTorch module and the TensorRT engine-based TorchScript.  \n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:25]: Please don't describe "model zoo" in the document, use "This tutorial shows how to convert the model of a release bundle to a TensorRT engine-based TorchScript for better performance and benchmark the difference between the PyTorch native model and the TensorRT engine-based TorchScript."

File: acceleration/TensorRT_inference_acceleration.ipynb (hunk 4 above)
Comment (binliunls) [reply acceleration/TensorRT_inference_acceleration.ipynb:89]: FYI, if directly run the tutorial in the ci/ci environment, it reports an torch-tensorrt related error, which means the torch-tensorrt directly installed by pip doesn't work in the environment.

Thanks,
Bin

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 10]
```diff
@@ -0,0 +1,930 @@
...
+   "source": [
+    "def get_parser(meta_config, inference_config):\n",
+    "    parser = monai.bundle.ConfigParser()\n",
+    "    parser.read_meta(meta_config)\n",
+    "    parser.read_config(inference_config)\n",
+    "    return parser\n",
+    "\n",
+    "\n",
+    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
+    "model_weight = glob.glob(os.path.join(\"./spleen_ct_segmentation\", \"*\", \"model.pt\"))[0]\n",
```
Comment (Nic-Ma) [inline acceleration/TensorRT_inference_acceleration.ipynb:236]: "./spleen_ct_segmentation/models/model.pt" may be enough? Why use glob.

File: acceleration/TensorRT_inference_acceleration.ipynb (hunk 9 above)
Comment (yiheng-wang-nv) [reply acceleration/TensorRT_inference_acceleration.ipynb:25]: convert the model of a release bundle >  convert the model of a **released** bundle

File: acceleration/TensorRT_inference_acceleration.ipynb [hunk 11]
```diff
@@ -0,0 +1,930 @@
+{
```
Comment (yiheng-wang-nv) [inline acceleration/TensorRT_inference_acceleration.ipynb:1]: <p>Suggest to attach the download script for convenience.</p><p><code></code><code>python</p><p>import os</p><p>import tempfile</p><p>from monai.apps import download_and_extract</p><p> </p><p>resource = "https://msd-for-monai.s3-us-west-2.amazonaws.com/Task09_Spleen.tar"</p><p>md5 = "410d4a301da4e5b2f6f86ec3ddba524e"</p><p> </p><p>directory = os.environ.get("MONAI_DATA_DIRECTORY")</p><p>root_dir = tempfile.mkdtemp() if directory is None else directory</p><p>compressed_file = os.path.join(root_dir, "Task09_Spleen.tar")</p><p>data_dir = os.path.join(root_dir, "Task09_Spleen")</p><p>if not os.path.exists(data_dir):</p><p>    download_and_extract(resource, compressed_file, root_dir, md5)</p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1215/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (yiheng-wang-nv) [inline acceleration/TensorRT_inference_acceleration.ipynb:1]: <blockquote>MONAI Docker whose version is greater than 1.2</blockquote><p> </p><p>should be &gt;= 1.2</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1215/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (yiheng-wang-nv) [inline acceleration/TensorRT_inference_acceleration.ipynb:1]: <p>my suggestion is to combine these three figure together, or unify the y-axis range (such as from 1.2 to 5.2) and put three figures into the same line.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1215/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (yiheng-wang-nv) [reply acceleration/TensorRT_inference_acceleration.ipynb:1]: Review NB seems has some issue here. This comment is for figures shown in the `Benchmark the model inference time` section

File: README.md [hunk 12]
```diff
@@ -205,6 +205,8 @@ This notebook is a quick demo for devices, run the Ignite trainer engine on CPU,
 Demonstrates the use of the `ThreadBuffer` class used to generate data batches during training in a separate thread.
 ##### [transform_speed](./acceleration/transform_speed.ipynb)
 Illustrate reading NIfTI files and test speed of different transforms on different devices.
+##### [TensorRT_inference_acceleration](./acceleration/TensorRT_inference_acceleration.ipynb)
+This note book shows how to use TensorRT to accelerate the model and achieve a better inference latency.
```
Comment (mingxin-zheng) [inline README.md:209]: note book -> notebook

File: acceleration/TensorRT_inference_acceleration.ipynb
Comment (mingxin-zheng) [inline acceleration/TensorRT_inference_acceleration.ipynb:1]: Can we use lower case for the filenames? For example
```
tensorrt_inference_acceleration.ipynb
```

**Comment by binliunls** [comment]
> Hi @binliunls ,
> 
> I see you use `ConfigParser` to handle a bundle config in multiple places of this PR, please update all of them to use `monai.bundle.ConfigWorkflow` instead. You can refer to: https://github.com/Project-MONAI/MONAI/blob/dev/tests/test_bundle_workflow.py#L204. And use `ConfigWorkflow initialize/run/finalize` to execute the inference. CC @yiheng-wang-nv for any other `BundleWorkflow` related questions.
> 
> Thanks.

After an offline synchronization, we think this is a very customer logic and will not use the ConfigWorkflow in this part.

Thanks,
Bin

File: acceleration/TensorRT_inference_acceleration.ipynb
Comment (binliunls) [reply acceleration/TensorRT_inference_acceleration.ipynb:1]: I saw that there are some tutorials also use the upper letter in their title like `.modules/TorchIO_MONAI_PyTorch_Lightning.ipynb`. @Nic-Ma @mingxin-zheng  Does this a requirement for the toolkit? If so, I can update it.

Thanks,
Bin

Comment (mingxin-zheng) [reply acceleration/TensorRT_inference_acceleration.ipynb:1]: It is not a hard requirement, but most of the tutorials are using lower case. Feel free to keep the filename as it is.

**Review by yiheng-wang-nv — APPROVED** [review]
my comments are all addressed by @binliunls , thanks and this PR looks good on my side.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update, looks good to me now.
